### ISSUE TITLE
entropy estimation

### ISSUE THREAD

**Issue by wensihan** [fab-jul/imgcomp-cvpr#4]
Hello, I am very confused of the part of context model in your code, what's the output of the context model represents? And what's the meaning of  pc_loss = beta * tf.maximum(H_soft - H_target, 0), the reason for setting H_target is for what? Thank you very much and looking forward to your reply~

**Comment by fab-jul** [comment]
The output of the context model is a conditional probability distribution.

`pc_loss = beta * tf.maximum(H_soft - H_target, 0)` is described on p. 4399 [in the paper](http://openaccess.thecvf.com/content_cvpr_2018/papers/Mentzer_Conditional_Probability_Models_CVPR_2018_paper.pdf).

**Comment by wensihan** [comment]
Then, if I don't use the H_target, it does not matter the performance, right?

**Comment by habibian** [comment]
In principle it should does, because entropy term force the autoencoder to
learn low entropy (compressible) codes for image.

**Comment by wensihan** [comment]
@habibian Thank you very much~ And by the way, do you have any idea of the realization of the adaptive arithmetic coding part?

**Comment by habibian** [comment]
Yes, you can use the context model to sequentially estimate the probability distribution symbol-by-symbol (as auto-regressive model decoding phase) and give it to the adaptive arithmetic coder as the current estimation of symbol probabilities.

**Comment by wensihan** [comment]
@habibian First, thank you again for your reply~ But there are still some confusion. We use the context model to sequentially estimate the probability distribution symbol-by-symbol, then we get a probability for each location in the feature map and with the probability we can use the adaptive arithmetic coding to get a final probability for each feature map as the seed, right? And then how can we decode the seed with adaptive arithmetic coder?

**Comment by fab-jul** [comment]
We updated the code to support arithmetic encoding to a bitstream. Please check out the updated [val.py:L161-L175](https://github.com/fab-jul/imgcomp-cvpr/blob/6058ace6aeecaf6dce4a6f4b7364a2e48461d511/code/val.py#L161-L175) and specifically [bit_counter.py](https://github.com/fab-jul/imgcomp-cvpr/blob/master/code/bit_counter.py), as well as the updated README.

**Comment by fab-jul** [comment]
Please check out #7
